packages/angular_devkit/core/src/jobs/simple-registry_spec.ts

```diff
@@ -0,0 +1,578 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+// tslint:disable:no-big-function no-non-null-assertion
+import { EMPTY, Observable, of, timer } from 'rxjs';
+import { map, take, toArray } from 'rxjs/operators';
+import { JobEvent, JobEventKind, JobHandlerContext, JobState } from './api';
+import { createJobHandler } from './create-job-handler';
+import { SimpleJobRegistry } from './simple-registry';
+import { SimpleScheduler } from './simple-scheduler';
+
+describe('SimpleJobRegistry', () => {
+  let registry: SimpleJobRegistry;
+  let scheduler: SimpleScheduler;
+
+  beforeEach(() => {
+    registry = new SimpleJobRegistry();
+    scheduler = new SimpleScheduler(registry);
+  });
+
+  it('works for a simple case', async () => {
+    registry.register(
+      'add', createJobHandler((arg: number[]) => arg.reduce((a, c) => a + c, 0)), {
+        argument: { items: { type: 'number' } },
+        output: { type: 'number' },
+      },
+    );
+
+    const sum = await (scheduler.schedule('add', [1, 2, 3, 4])).output.toPromise();
+    expect(sum).toBe(10);
+  });
+
+  it('calls jobs in parallel', async () => {
+    let started = 0;
+    let finished = 0;
+
+    registry.register(
+      'add',
+      createJobHandler((argument: number[]) => {
+        started++;
+
+        return new Promise<number>(
+          resolve => setTimeout(() => {
+            finished++;
+            resolve(argument.reduce((a, c) => a + c, 0));
+          }, 10),
+        );
+      }),
+      {
+        argument: { items: { type: 'number' } },
+        output: { type: 'number' },
+      },
+    );
+
+    const job1 = scheduler.schedule('add', [1, 2, 3, 4]);
+    const job2 = scheduler.schedule('add', [1, 2, 3, 4, 5]);
+    expect(started).toBe(0);
+
+    const p1 = job1.output.toPromise();
+    expect(started).toBe(1);
+
+    const p2 = job2.output.toPromise();
+    expect(started).toBe(2);
+    expect(finished).toBe(0);
+
+    const [sum, sum2] = await Promise.all([p1, p2]);
+    expect(started).toBe(2);
+    expect(finished).toBe(2);
+
+    expect(sum).toBe(10);
+    expect(sum2).toBe(15);
+  });
+
+  it('validates arguments', async () => {
+    registry.register(
+      'add', createJobHandler((arg: number[]) => arg.reduce((a, c) => a + c, 0)), {
+        argument: { items: { type: 'number' } },
+        output: { type: 'number' },
+      },
+    );
+
+    await (scheduler.schedule('add', [1, 2, 3, 4])).output.toPromise();
+    try {
+      await (scheduler.schedule('add', ['1', 2, 3, 4])).output.toPromise();
+      expect(true).toBe(false);
+    } catch {}
+  });
+
+  it('validates outputs', async () => {
+    registry.register(
+      'add', createJobHandler(() => 'hello world'), {
+        output: { type: 'number' },
+      },
+    );
+
+    try {
+      await (scheduler.schedule('add', [1, 2, 3, 4])).output.toPromise();
+      expect(true).toBe(false);
+    } catch {}
+  });
+
+  it('works with dependencies', async () => {
```

Comment:
We don't, but I feel preventing dependency cycles is out of scope.

There are two ways to do a cycle in the current API: either you depend on yourself but have a strategy which is currently waiting (e.g. a `serialize()` job calling itself), or you have a shared `Job<>` instance and you wait on it. That second case is breaking a convention here that jobs should not share memory with each others.

The first case is more interesting, but I'd say it's similar to this code:

```javascript
let s1, o1 = new Observable(s => { s1 = s; });
let s2, o2 = new Observable(s => { s2 = s; });

s2.subscribe(undefined, undefined, () => s1.complete());
s1.subscribe(undefined, undefined, () => s2.complete());

await o1.toPromise();  // Deadlocked.
```

I don't know of any way to test that, and I feel it's out of scope of this library. The user should be responsible (we can provide utility functions for it).

Reply:
One way to avoid the cyclical graph issue is to remove dependencies as a concept from this level of the architecture.
There are already several ways to model a dependency graph without them.  For example, a consumer could create a hierarchy of Jobs by having the Jobs themselves create and manage their children (which would fit architect quite well).  Or a consumer can directly manage the start of each individual job (which based on the current design, they are already required to do).  The later would be especially viable if the dependency model is already explicit in the consumer's design (schematic tasks).

Reply:
As long as `serialize()` exists as a strategy you can lock yourself. I don't know if removing dependencies from the API layer is good enough. Also, dependencies don't lock you in; passing a dependency to your own job through shared memory does. This is a breach of general contract though; the reason every value is serializable is to enforce that your job can be run in a separate process.